src/USER-PHONON/third_order.cpp

```diff
@@ -242,26 +257,30 @@ void ThirdOrder::calculateMatrix()
   bigint *gm = groupmap;
   double **f = atom->f;
 
-  double *dynmat = new double[3*dynlen];
-  double *fdynmat = new double[3*dynlen];
-  memset(&dynmat[0],0,dynlen*sizeof(double));
-  memset(&fdynmat[0],0,dynlen*sizeof(double));
+  double *dynmat = new double[dynlenb];
+  double *fdynmat = new double[dynlenb];
+  memset(&dynmat[0],0,dynlenb*sizeof(double));
+  memset(&fdynmat[0],0,dynlenb*sizeof(double));
 
   if (comm->me == 0 && screen) {
     fprintf(screen,"Calculating Third Order ...\n");
     fprintf(screen,"  Total # of atoms = " BIGINT_FORMAT "\n", natoms);
     fprintf(screen,"  Atoms in group = " BIGINT_FORMAT "\n", gcount);
     fprintf(screen,"  Total third order elements = "
-            BIGINT_FORMAT "\n", (dynlen*dynlen*dynlen) );
+            BIGINT_FORMAT "\n", (dynlen*dynlenb*dynlenb) );
   }
 
   update->nsteps = 0;
   int prog = 0;
   for (bigint i=1; i<=natoms; i++){
     local_idx = atom->map(i);
+    if (gm[i-1] < 0)
+      continue;
     for (int alpha=0; alpha<3; alpha++){
       for (bigint j=1; j<=natoms; j++){
         local_jdx = atom->map(j);
+        if (gm[j-1] < 0 && !folded)
```

Comment:
This might be a tall order, but I doubt it. I would like to check if atom j is a neighbor of atom i. It is needless to continue in the for loop otherwise and this addition would heavily reduce the number of force compute calls.

I am currently trying to reading through compute coord/atom to see how the compute forms its own neighbor list, thanks to a great suggestion by @athomps. Any additional advice is welcome.

Reply:
@charlessievers i would rather look at commands like `create_bonds` or `delete_atoms` for how to request and build a neighborlist from inside a command. This can be a full or a half neighbor list.

Short of using that, there also is the option to do spatial binning aka cell lists (if for some reason neighbor lists are not practical). You can globally bin all atoms (an O(N) process) and then only have to compute the closest distance between two bins to decide whether all pairs of atoms formed between the two bins are outside of a given cutoff and can be skipped. that would be a significant savings although not as much as using neighbor lists.

Reply:
I will look at those, thank you!

Reply:
Bear with me on my next question. Does ilist always contain every owned local non-ghost id?

Reply:
> Bear with me on my next question. Does ilist always contain every owned local non-ghost id?

most of the time. not when atoms are explicitly excluded from the neighbor list. you can easily check for that with comparing `inum` to `atom->nlocal`

Reply:
Excellent, first time really looking into LAMMPS neighbor lists so I am sorry for the basic questions.

Reply:
also ilist does not contain ids (aka tags) but local indices, i.e. 0 to atom->nlocal. if you want the id you need `atom->tag[ilist[j] & NEIGHMASK]`. also you need to watch out for special_bond bits, hence the application of NEIGHMASK.

Reply:
I assume, based on what you just said, there is no trivial atom->map() equivalent to go from tag to ilist index "ii" or jlist index "jj"

Reply:
I thought about this a while ago, but somehow got stuck on the neighbors-of-neighbors problem: moving one atom can change the force on an atom two neighbors over by chaning \rho of the inbetween one. I think only atoms at 2* force cutoff can be safely ignored?

Reply:
That is correct. The necessary but not sufficient condition for non-zero dFj/dRi is:
    there exists atom k s.t. i and j are neighbors of k, where k is always a neighbor of itself.
A fairly straightforward way to find all non-zero dFj/dRi is to first build a full neighbor list, then loop over atoms k, then do a double loop over neighbors of k, adding to a new list storing all ordered (i,j) pairs that satisfy the above criterion.

Reply:
> I assume, based on what you just said, there is no trivial atom->map() equivalent to go from tag to ilist index "ii".

why would you need that? is there anything that forces you in looping over tags in order? you can always get the tag from atom->tag[i] or atom->tag[j].

Reply:
> I thought about this a while ago, but somehow got stuck on the neighbors-of-neighbors problem: moving one atom can change the force on an atom two neighbors over by chaning \rho of the inbetween one. I think only atoms at 2* force cutoff can be safely ignored?

unless you have long-range electrostatics or a model with dihedrals and a short pairwise cutoff (in which case you should get a warning to increase the communication cutoff).

Reply:
Just pushed a **draft** of my solution. I only search through first neighbors. I realize that I need to loop over their neighbors as well, but I figured it would be nice to have a draft to work off of.

Reply:
New draft produced a neighbor/neighbor_of_neighbor list (extended neighbor list). This list is indexed by atom tag and yields a pointer to an array that contains the atom tags of the extended neighbors. Next step for me is to make this depend on the third order's group map member.